src/feature/hs/hs_client.c

```diff
@@ -956,6 +956,37 @@ client_get_random_intro(const ed25519_public_key_t *service_pk)
   return ei;
 }
 
+/** Return true iff all intro points for the given service have timed out. */
```

Comment:
It's a bit weird commit structure here where the first commit is about intro, the second comimt is rend, and the third commit is again about intro and also editing the firstr commit. I had to rebase it and squash the intro stuff into a single commit to be able to review it effectively.

Reply:
Right so the code is a bit messy to follow but the structure is a commit per error code.